astropy/wcs/utils.py

```diff
@@ -632,3 +632,213 @@ def pixel_to_skycoord(xp, yp, wcs, origin=0, mode='all', cls=None):
+
+    # empty wcs to fill in with fit values
+    wcs = celestial_frame_to_wcs(frame=world_coords.frame,
+                                 projection=projection)
+
+    # set pixel_shape to span of input points
+    wcs.pixel_shape = (max(xp)-min(xp), max(yp)-min(yp))
+
+    # determine CRVAL from input
+    close = lambda l, p: p[np.where(np.abs(l) == min(np.abs(l)))[0][0]]
+    if str(proj_point) == 'center':  # use center of input points
+        wcs.wcs.crval = ((max(lon)+min(lon))/2., (max(lat)+min(lat))/2.)
+        wcs.wcs.crpix = ((max(xp)+min(xp))/2., (max(yp)+min(yp))/2.)
+    elif proj_point is not None:  # convert units, initial guess for crpix
+        proj_point.transform_to(world_coords)
+        wcs.wcs.crval = (proj_point.data.lon.deg, proj_point.data.lat.deg)
+        wcs.wcs.crpix = (close(lon-wcs.wcs.crval[0], xp),
+                         close(lon-wcs.wcs.crval[1], yp))
+
+    # fit linear terms, assign to wcs
+    p0 = np.concatenate((wcs.wcs.pc.flatten(), wcs.wcs.crpix.flatten()))
+    fit = least_squares(_linear_wcs_fit, p0,
+                        args=(lon, lat, xp, yp, wcs))
+    wcs.wcs.crpix = np.array(fit.x[4:6])
+    wcs.wcs.pc = np.array(fit.x[0:4].reshape((2, 2)))
+
+    # fit SIP, if specified. 
+    if sip_distortion:
+        wcs.wcs.ctype = [x + '-SIP' for x in wcs.wcs.ctype]
+        coef_names = ['{0}_{1}'.format(i, j) for i in range(degree+1)
+                      for j in range(degree+1) if (i+j) < (degree+1) and
+                      (i+j) > 1]
+        p0 = np.concatenate((np.array(wcs.wcs.crpix), wcs.wcs.pc.flatten(),
+                             np.zeros(2*len(coef_names))))
+
+        fit_reverse = least_squares(_sip_fit, p0, args=(lon, lat, xp, yp, 
+                                    wcs, degree, coef_names,'reverse'))
+
+        fit = least_squares(_sip_fit, p0,
+                            args=(lon, lat, xp, yp, wcs, degree, coef_names))
+
+        # put fit values in wcs
+        wcs.wcs.pc = fit.x[2:6].reshape((2, 2))
+        wcs.wcs.crpix = fit.x[0:2]
+        coef_fit = (list(fit.x[6:6+len(coef_names)]),
+                    list(fit.x[6+len(coef_names):]))
+        coef_fit_reverse = (list(fit_reverse.x[6:6+len(coef_names)]),
+                    list(fit_reverse.x[6+len(coef_names):]))
+        
+        a_vals = np.zeros((degree+1, degree+1))
+        b_vals = np.zeros((degree+1, degree+1))
+        ap_vals = np.zeros((degree+1, degree+1))
+        bp_vals = np.zeros((degree+1, degree+1))
+        for coef_name in coef_names:
+            a_vals[int(coef_name[0])][int(coef_name[2])] = coef_fit[0].pop(0)
+            b_vals[int(coef_name[0])][int(coef_name[2])] = coef_fit[1].pop(0)
+            ap_vals[int(coef_name[0])][int(coef_name[2])] = coef_fit_reverse[0].pop(0)
+            bp_vals[int(coef_name[0])][int(coef_name[2])] = coef_fit_reverse[1].pop(0)
+        wcs.sip = Sip(a_vals, b_vals, ap_vals, bp_vals, wcs.wcs.crpix)
+
+    return wcs
```

Comment:
@shannnnnyyy can correct me if I'm mis-remembering, but I *think* this fitting algorithm was pulled from a specific known-good approach for some HST dataset (maybe WFC3?).  It that case it might be good to know if there's some experience that suggests the approach in here now is better for some pathological cases...? But if not I agree @nden's suggestion seems more straightforward and stable.

Reply:
Another option, given the impending 4.0 feature freeze, is to merge this using the current algorithm but implement @nden's solution as a follow-on PR (which then might end up slipping to 4.1) which is triggered via a specific keyword or something like that.